platform-sdk/docs/components/componentFramework.md

```diff
@@ -0,0 +1,218 @@
+# Component Framework
+
+## Goals
+
+The goal of this framework is to provide a structure for defining software components in which:
+
+1. Components can be defined in isolation, each with their own unit of business logic.
+2. Components can be connected (or wired) together with other components to form a data pipeline.
+3. Each component allows its threading model to be changed individually without modifying the business
+   logic.
+
+These goals allow software to be well organized, flexible, and easily tuned for performance.
+
+## Overview
+
+At the most basic level, a component in this framework is composed of a task scheduler, business logic, and wires. Each
+component has a unit of business logic that performs some action on a piece of data. The component receives the data on
+an input wire, performs the action, and optionally produces data on an output wire. The components are connected, or
+"soldered", together by joining input and output wires of the same data type to form a data pipeline.
+
+The concurrency of the component is determined by the task scheduler type. Each component has a task scheduler which
+combines the data received on the input wire with the business logic operation to form a task. This task is then
+scheduled for execution according to the task scheduler type and implementation. Task schedulers are described in more
+detail [below](#TaskSchedulers).
+
+Here is a picture of a simple component with one input wire of type `Integer`, and one output wire of type `String`:
+
+![SimpleComponent](step4.drawio.svg)
+
+### Business Logic
+
+Business logic is defined in one or more classes that work together and accept data, perform an
+operation, and produce optional output. This framework allows business logic to be defined completely independently
+of the threading model and other business logic. Each unit of business logic does not know where its input comes from
+or who consumes its output. The threading is handled by its task scheduler, and the connection of inputs and outputs is
+handled externally. Changes to concurrency and wiring have no impact on the business logic.
+
+### Wires
+
+Wires are the construct on which data flows between components. Each component can have one or more input wires
+(restrictions apply based on the concurrency of the task scheduler) and one or more output wires. Components that do not
+produce any data have an output of type `Void`. Wires are strongly typed to the data they transport. Wires of the
+same type can be "soldered" or connected to determine the flow of data between components.
+
+### Task Schedulers
+
+Task schedulers are responsible for several things, but the most important job is taking data received on the input
+wire, combining it with the business logic operation to be performed on that data to create a task, and scheduling that
```

Comment:
>Checking my understanding, the logical queue in the diagram below stores the "tasks," which have been added (scheduled) by the previous component in the chain.

Yes, though some implementations do not store the task at all. Instead, they keep a counter of the number of tasks that have not completed execution and block until the counter is below the threshold. This is why it is a "logical" queue.

>...as a final activity for the task, it should schedule the next component task

Not quite. If the thread trying to schedule a task is blocked because there is backpressure, it will remain blocked until there is room to schedule the task. The thread that completes execution of a task that results in there being enough room for the next task to be scheduled does not schedule the next task. That is done by the thread that was blocked. This is important, because in order for backpressure to work, the thread that is scheduling tasks (and executing tasks earlier in the pipeline) must be stopped from creating even more tasks. The purpose is to stop even more tasks from piling up.